### ISSUE TITLE
placeholderArgs: don't collapse missing indexed placeholders

### ISSUE THREAD

**Issue by plexigras** [wcjohnson/lightscript#21]
```js
(=> _2)(0, 1, 2, 3)
```

it gets compiled to

```js
(_arg => _arg)(0, 1, 2, 3);
```

i would prefer it to compile to

```js
((_arg0, _arg1, _arg2) => _arg2)(0, 1, 2, 3);
```

this is similar to https://github.com/citycide/babel-plugin-partial-application/issues/7

**Comment by plexigras** [comment]
i have no idea what im doing but something like this might work

```diff
  // Formulate args
  args = []
-  let prevIndex = undefined, restArg = undefined
+  let prevIndex = -1, restArg = undefined
  for elem phPath in sortedPhPaths:
    { node } = phPath
    if phPath.parentPath?.node.type == "SpreadElement":
      if not restArg:
        now restArg = fnPath.scope.generateUidIdentifier("arg")
        args.push(t.restElement(restArg))
      phPath.replaceWith(restArg)
    elif node.index? and node.index == prevIndex:
      // Repeated use of an index = same arg.
      phPath.replaceWith(args[args.length - 1])
    else:
-       args.push(fnPath.scope.generateUidIdentifier("arg"))
-       now prevIndex = node.index
+       while prevIndex < node.index:
+         args.push(fnPath.scope.generateUidIdentifier("arg"))
+         now prevIndex++
      phPath.replaceWith(args[args.length - 1])

  // Lift args to function level
  fnNode = fnPath.node
  fnNode.params = args
  fnPath.replaceWith(fnNode)
```
https://github.com/wcjohnson/babel-plugin-lightscript/blob/8ccd7166deabc485ca1835c533d7db436be0a5c8/src/placeholders.lsc#L37-L58

**Comment by wcjohnson** [comment]
Agreed. I think also I will disable mixing non-indexed and indexed placeholders in the same function.

**Comment by plexigras** [comment]
I tried it out and it actually works :)

**Comment by plexigras** [comment]
Should i make a PR?

I can also add this to disable mixing non-indexed and indexed placeholders.

```js
if ixPhPaths.length and nixPhPaths.length:
  throw fnPath.buildCodeFrameError("Indexed and non-indexed placeholders can not be used in the same function.")
```

https://github.com/wcjohnson/babel-plugin-lightscript/blob/8ccd7166deabc485ca1835c533d7db436be0a5c8/src/placeholders.lsc#L36

**Comment by wcjohnson** [comment]
PR is welcome. Make sure you add/update the unit tests: https://github.com/wcjohnson/babel-plugin-lightscript/tree/8ccd7166deabc485ca1835c533d7db436be0a5c8/test/fixtures/placeholder-args

**Comment by wcjohnson** [comment]
Actually this might be a good opportunity to fix https://github.com/wcjohnson/lightscript/issues/20 as well (no guaranteed stability on array.sort)

**Comment by plexigras** [comment]
how can i generate the uinit test automatically?

**Comment by plexigras** [comment]
i don't know how to make the pull request but here you go

https://github.com/wcjohnson/babel-plugin-lightscript/compare/8ccd716...plexigras:fix-collapsing-placeholders
https://github.com/wcjohnson/babel-plugin-lightscript/compare/8ccd716...plexigras:fix-disable-placeholder-mixing

**Comment by wcjohnson** [comment]
OK, I've done some work on this, which amounts to a slight redesign of the placeholder feature.

- Mixing indexed placeholders and non-indexed placeholders is now illegal.

- Non-indexed placeholders work like Kotlin `it` and always refer to the first arg. E.g.:
```js
-> f(_, _)
```
yields
```js
function(arg0) { return f(arg0, arg0); }
```
instead of
```js
function(arg0, arg1) { return f(arg0, arg1) }
```

- Indexed placeholders are no longer collapsed and refer to the specific (zero-indexed) arg. So e.g.
```js
-> _2
```
yields
```js
function(arg0, arg1, arg2) { return arg2; }
```

- Spread placeholders can be used in both cases and generate a rest arg at the end of the arg list.

Changes implemented here: https://github.com/wcjohnson/babel-plugin-lightscript/commit/6b9f6b700314e94f94221b640854f7defc718816

**Comment by wcjohnson** [comment]
Also fixes https://github.com/wcjohnson/lightscript/issues/20

**Comment by wcjohnson** [comment]
Implemented/fixed in 2.3.0-alpha.2
